### PR TITLE
SOA test

### LINKED ISSUE #49: [tests] Test for SOA bot
We need a test in `tests/` for SOA bot. It can look similar to the other bot tests, but should be in its own file.

### PR DESCRIPTION
**Pull request by wwongkamjan**
resolves #49 

I do think that the SOA test itself also needs some tests :) 

At least for now, we could check if the tests are all reasonable and if you'd like me to add any other test case, please let me know.

### CODE REVIEW

File: tests/bot_tests/SOA_test.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+        self.test_play()
+        self.test_stance()
+        self.test_get_proposals()
+        self.test_get_best_orders()
+        self.test_gen_messages()
+
+    def test_play(self):
+        game_play = GamePlay(None, [RandomProposerBot, RandomProposerBot, SmartOrderAccepterBot], 3, True)
+        msgs, done = game_play.step()
+        print(msgs)
```
Comment (trigaten) [inline tests/bot_tests/SOA_test.py:30]: remove print statement

File: tests/bot_tests/SOA_test.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+    def test_stance(self):
+        # score-based
+        SOA_bot = SmartOrderAccepterBot
+        bot_instances = [RandomProposerBot, RandomProposerBot, SOA_bot, RandomProposerBot ]
+        game_play = GamePlay(None, bot_instances, 3, True)
+        game_play.game.set_centers('AUSTRIA', ['VIE','TRI','BUD'], reset=True)
+        game_play.game.set_centers('ENGLAND', ['LON'])
+        game_play.game.set_centers('GERMANY', ['MUN', 'KIE', 'BER','BEL'])
+        game_play.game.set_centers(SOA_bot.power, ['PAR','BRE', 'MAR'])
```
Comment (trigaten) [inline tests/bot_tests/SOA_test.py:42]: How is `SOA_bot.power` being used without an instantiated instance of `SOA_bot.power`?

File: tests/bot_tests/SOA_test.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        msgs, done = game_play.step()
+        SOA_bot_stance = SOA_bot.stance.get_stance()[SOA_bot.power]
```
Comment (trigaten) [inline tests/bot_tests/SOA_test.py:44]: same here

File: tests/bot_tests/SOA_test.py (hunk 2 above)
Comment (wwongkamjan) [reply tests/bot_tests/SOA_test.py:42]: Right. that's silly mistake xD

**Comment by wwongkamjan** [comment]
I instantiate all the bots in use for tests. Please review it again. Thanks!

**Comment by trigaten** [comment]
@wwongkamjan Just to make sure, have you run this test on TACC, and does it pass?

File: tests/bot_tests/SOA_test.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        assert SOA_bot_stance['GERMANY'] == -1, "Negative stance error"
+        assert SOA_bot_stance['AUSTRIA'] == 0 , "Neutral stance error"
+
+        # to do: add action-based stance test
+
+    def test_get_proposals(self):
+        # proposal messages -> proposal dict {power_name: a list of proposal orders}
+        # valid moves and power units must belong to SOA
+        game = Game()
+        SOA_bot = SmartOrderAccepterBot('FRANCE', game)
```
Comment (trigaten) [inline tests/bot_tests/SOA_test.py:57]: remove capitalization in this variable name to avoid confusion with classes

Comment (trigaten) [reply tests/bot_tests/SOA_test.py:57]: generally variables names shouldn't be capitalized

**Comment by wwongkamjan** [comment]
> @wwongkamjan Just to make sure, have you run this test on TACC, and does it pass?

Have not. I think I have to wait until SOA bot done to run these test cases

**Comment by trigaten** [comment]
I do not want to merge a test that has not been run. What needs to be changed in the bot?

**Comment by trigaten** [comment]
@wwongkamjan Since the bot is able to run on TACC, are there any updates on this?

File: tests/bot_tests/SOA_test.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+                assert unit_order in soa_power_units, "unit in " + order + " does not belong to SOA's power (" + soa_bot.power_name + ")"
+                assert order in possible_orders[unit_order], order + " is not possible in this current phase of a game for SOA's power (" + soa_bot.power_name + ")"
+
+
+    def test_get_best_orders(self):
+        # proposal -> gen state value check if SOA select the best proposal 
+        
+        #run test for n times
+        n=100
+        orders = yield self.brain.get_orders(self.game, self.power_name)
```
Comment (trigaten) [inline tests/bot_tests/SOA_test.py:106]: Be careful here... what is self? I don't think its a bot.

**Pull request closed without merging**